### PR TITLE
feat: add to_parquet_row_groups

### PR DESCRIPTION
**Pull request by zbilodea**

### CODE REVIEW

File: src/awkward/operations/ak_to_parquet.py
Comment (jpivarski) [inline src/awkward/operations/ak_to_parquet.py]:
```python
layout = ak.operations.ak_to_layout._impl(data, ...)
```

will convert `data` into an Awkward Array, and Awkward Arrays are eagerly in memory all at once. `ak.to_parquet_row_groups` is supposed to take an iterator as `data`, so this will iterate over it, pulling it all into memory, which is not what you want.

In the case of `ak.to_parquet_row_groups`, you'll want to do a Python for loop over `data` and call `ak_to_layout` on each item individually.

File: src/awkward/operations/ak_to_parquet.py
Comment (jpivarski) [inline src/awkward/operations/ak_to_parquet.py]: `iter` is a builtin Python function name, one that you might want to use in the implementation (if it doesn't fit well into a single `for` loop). By using it as the boolean back-door, you lose access to that function. (You can get it back with `__builtins__.iter`, but that's deliberately making things difficult.)

```python
>>> def generator():
...     for _ in range(5):
...         print("making another one")
...         yield "another one"
... 
>>> it = iter(generator())
>>> next(it)
making another one
'another one'
>>> next(it)
making another one
'another one'
>>> next(it)
making another one
'another one'
>>> next(it)
making another one
'another one'
>>> next(it)
making another one
'another one'
>>> next(it)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

**Review by jpivarski — CHANGES_REQUESTED** [review]
You might not need to define the `batch_iterator` function. (Is it a copy of what was in Awkward 1.x? This is a different context.) You're already iterating over batches in ak_to_parquet.py lines 408‒435. I haven't looked closely enough at it to point out whether something is superfluous, but it looks like it might be. It's worth focusing on that a bit more.

I wish there was a way to write a test that would check to see that all the batches are not loaded into memory at once. Maybe you could try it on a manual test, outside of CI, with a dataset that is too large to fit in memory at once. Putting (temporary!) print statements in some places would also let you check the temporal order of things.

File: src/awkward/operations/ak_to_parquet.py
Comment (zbilodea) [reply src/awkward/operations/ak_to_parquet.py]: Okay! Now calling to_layout for each iteration - I printed each new layout to check that they only have data from one array and it looks good, but I'll test it on a large dataset to be sure

Comment (zbilodea) [reply src/awkward/operations/ak_to_parquet.py]: I made a test and was able to write file larger than my laptop's memory!

Screenshot 2024-01-29 at 11 54 11

Comment (jpivarski) [reply src/awkward/operations/ak_to_parquet.py]: Also, the output file is compressed, whereas the data in memory were not compressed.

File: src/awkward/operations/ak_to_parquet.py
Comment (jpivarski) [inline src/awkward/operations/ak_to_parquet.py]: Python makes a distinction between an iterator (something that you can call `next` on to get another item) and an iterable (something that you can call `iter` on to get an iterator). The interface should work for iterables, so `iter(data)` will be needed at some point.

**Review by jpivarski — APPROVED** [review]
This is good in that it works, but the code is a bit repetitive and the control flow is hard to follow. I'll just make a few edits after this review: don't merge until I'm done.

**Comment by jpivarski** [comment]
This refactoring reduces three repetitions of `layout = ...` followed by `table = ...` into one by defining a temporary function, and it simplifies the `while ... try ... except StopIteration` logic with a `for` loop. It also adds guards to prevent users from making some likely mistakes.

The `for` loop takes advantage of the distinction between iterables (such as lists, array, and iterators) and iterators: when the first items in an iterator have been consumed with `next`, a `for` loop over the iterator does not include the already-consumed items—Python does not go back to the beginning.

```python
>>> some_list = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"]
>>> next(some_list)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'list' object is not an iterator
```

```python
>>> iterator = iter(some_list)
>>> next(iterator)
'one'
>>> next(iterator)
'two'
>>> next(iterator)
'three'
>>> for item in iterator:
...     print(item)
... 
four
five
six
seven
eight
nine
ten
```

but

```python
>>> for item in some_list:
...     print(item)
... 
one
two
three
four
five
six
seven
eight
nine
ten
```

I think this PR is ready for merging, after doing an "Update branch." Look it over and decide if you want to make any counter-edits. If so, ping me again for another review. If not, go ahead and merge it.

File: src/awkward/operations/ak_to_parquet.py
Comment (jpivarski) [inline src/awkward/operations/ak_to_parquet.py]: This is the main thing that I wanted to simplify, for future maintainability.

**Pull request merged**
